vrx_gz/src/PerceptionScoringPlugin.cc

```diff
@@ -0,0 +1,518 @@
+      duration = durationElement->Get<double>();
+    }
+
+    // Parse the object type.
+    if (!objectElem->HasElement("type"))
+    {
+      gzerr << "PerceptionScoringPlugin: Unable to find <type> in object.\n";
+      objectElem = objectElem->GetNextElement("object");
+      continue;
+    }
+    sdf::ElementPtr typeElement = objectElem->GetElement("type");
+    std::string type = typeElement->Get<std::string>();
+
+    // Parse the object name - this is what must be matched id success.
+    if (!objectElem->HasElement("name"))
+    {
+      gzerr << "PerceptionScoringPlugin: Unable to find <name> in object.\n";
+      objectElem = objectElem->GetNextElement("object");
+      continue;
+    }
+    sdf::ElementPtr nameElement = objectElem->GetElement("name");
+    std::string name = nameElement->Get<std::string>();
+
+    // Parse the object pose
+    if (!objectElem->HasElement("pose"))
+    {
+      gzerr << "PerceptionScoringPlugin: Unable to find <pose> in object.\n";
+      objectElem = objectElem->GetNextElement("object");
+      continue;
+    }
+    sdf::ElementPtr poseElement = objectElem->GetElement("pose");
+    math::Pose3d pose = poseElement->Get<math::Pose3d>();
+
+    // Add the object to the collection.
+    PerceptionObject obj(time, duration, type, name, pose, _ecm);
+    gzdbg << "New object: " << std::endl << obj << std::endl;
+    this->objects.push_back(std::move(obj));
+
+    objectElem = objectElem->GetNextElement("object");
+  }
+
+  // Optional: <frame>.
+  if (this->sdf->HasElement("frame"))
+    this->frameName = this->sdf->Get<std::string>("frame");
+
+  // Optional: <landmark_topic>.
+  this->objectTopic = "/vrx/perception/landmark";
+  if (this->sdf->HasElement("landmark_topic"))
+  {
+    this->objectTopic =
+      this->sdf->GetElement("landmark_topic")->Get<std::string>();
+  }
+
+  return true;
+}
+
+//////////////////////////////////////////////////
+void PerceptionScoringPlugin::Implementation::OnAttempt(const msgs::Pose &_msg)
+{
+  std::lock_guard<std::mutex> lock(this->mutex);
+  this->requests.push_back(_msg);
+}
+
+//////////////////////////////////////////////////
+void PerceptionScoringPlugin::Implementation::ProcessAttempts(
+  sim::EntityComponentManager &_ecm)
+{
+  std::lock_guard<std::mutex> lock(this->mutex);
+
+  for (auto &_msg : this->requests)
+  {
+    // Only accept an attempt if there are any in the attempt balance.
+    if (this->attemptBal == 0)
+    {
+      gzwarn << "PerceptionScoring: Attempt Balance is 0, no attempts currently"
+             << " allowed. Ignoring." << std::endl;
+
+      this->requests.clear();
+      return;
+    }
+
+    // In the ROS side, the object type is reported in the "frame_id" field
+    // of the header.
+    std::string typeReported;
+    for (auto i = 0u; i < _msg.header().data_size(); ++i)
+    {
+      if (_msg.header().data(i).key() == "frame_id")
+        typeReported = _msg.header().data(i).value(0);
```

Comment:
`break` as well?

Reply:
Added in [950eb68](https://github.com/osrf/vrx/pull/525/commits/950eb68cb21b687e12d0b8146b8e74aa7cfe9dbd).